### PR TITLE
Youtube: Move video data fetching to a different function

### PR DESCRIPTION
**Pull request by mia-pi-git**
Requested by Kris for their plugin for the Studio. Partial credit to them for re-writing some of the code.

### CODE REVIEW

File: server/chat-plugins/youtube.ts [hunk 1]
```diff
@@ -129,6 +143,27 @@ export class YoutubeInterface {
...
+			id,
+			date: new Date(video.snippet.publishedAt).toString(),
+			description: video.snippet.description,
+			channelTitle: video.snippet.channelTitle,
+			channelUrl: video.snippet.channelId,
+			views: video.statistics.viewCount,
+			thumbnail: video.snippet.thumbnails.default.url,
+			likes: video.statistics.likeCount,
+			dislikes: video.statistics.dislikeCount,
+		} as VideoData;
```
Comment (Zarel) [inline server/chat-plugins/youtube.ts:165]: Is the `as VideoData` necessary? It can cover up bugs and should probably be handled individually.

Comment (Zarel) [inline server/chat-plugins/youtube.ts:165]:
```suggestion
		};
```

File: server/chat-plugins/youtube.ts [hunk 2]
```diff
@@ -129,6 +143,27 @@ export class YoutubeInterface {
...
+		const id = this.getId(link);
+		const raw = await Net(`${ROOT}videos`).get({
+			query: {part: 'snippet,statistics', id, key: Config.youtubeKey},
+		});
+		const res = JSON.parse(raw);
+		if (!res || !res.items || res.items.length < 1) return null;
+		const video = res.items[0];
+		return {
+			title: video.snippet.title,
+			id,
```
Comment (Zarel) [inline server/chat-plugins/youtube.ts:156]:
```suggestion
			id as ID,
```

Comment (KrisXV) [reply server/chat-plugins/youtube.ts]: This isnt actually an ID, it's a video ID (which has hyphens etc in it)

Comment (KrisXV) [reply server/chat-plugins/youtube.ts]: So hardcasting isnt good

Comment (KrisXV) [reply server/chat-plugins/youtube.ts]: im confused why it was labeled that in the interface though

File: server/chat-plugins/youtube.ts [hunk 3]
```diff
@@ -25,6 +25,19 @@ interface ChannelEntry {
 	category?: string;
 }
 
+interface VideoData {
+	id: ID;
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:29]:
```suggestion
	id: string;
```

File: server/chat-plugins/youtube.ts [hunk 4]
```diff
@@ -129,6 +143,27 @@ export class YoutubeInterface {
...
+		const id = this.getId(link);
+		const raw = await Net(`${ROOT}videos`).get({
+			query: {part: 'snippet,statistics', id, key: Config.youtubeKey},
+		});
+		const res = JSON.parse(raw);
+		if (!res || !res.items || res.items.length < 1) return null;
+		const video = res.items[0];
+		return {
+			title: video.snippet.title,
+			id as ID,
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:156]:
```suggestion
			id,
```

File: server/chat-plugins/youtube.ts (hunk 2 above)
Comment (Zarel) [reply server/chat-plugins/youtube.ts:156]: It's an easy mistake to make; it was labeled "ID" after all.

**Comment by Zarel** [comment]
@mia-pi-git anyway this is why your hardcast is bad; it covered up the bug of `id` not being an ID.

File: server/chat-plugins/youtube.ts (hunk 2 above)
Comment (KrisXV) [reply server/chat-plugins/youtube.ts:156]: ya i wasnt faulting anyone, just saying

**Pull request merged**